src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/FileBasedProgramsProjectSystem.cs

```diff
@@ -0,0 +1,143 @@
+        LanguageServerWorkspaceFactory workspaceFactory,
+        IFileChangeWatcher fileChangeWatcher,
+        IGlobalOptionService globalOptionService,
+        ILoggerFactory loggerFactory,
+        IAsynchronousOperationListenerProvider listenerProvider,
+        ProjectLoadTelemetryReporter projectLoadTelemetry,
+        ServerConfigurationFactory serverConfigurationFactory,
+        BinlogNamer binlogNamer)
+            : base(
+                workspaceFactory.FileBasedProgramsProjectFactory,
+                workspaceFactory.TargetFrameworkManager,
+                workspaceFactory.ProjectSystemHostInfo,
+                fileChangeWatcher,
+                globalOptionService,
+                loggerFactory,
+                listenerProvider,
+                projectLoadTelemetry,
+                serverConfigurationFactory,
+                binlogNamer)
+    {
+        _lspServices = lspServices;
+        _logger = loggerFactory.CreateLogger<FileBasedProgramsProjectSystem>();
+        _metadataAsSourceFileService = metadataAsSourceFileService;
+    }
+
+    public Workspace Workspace => ProjectFactory.Workspace;
+
+    public async Task<TextDocument?> AddMiscellaneousDocumentAsync(DocumentUri uri, SourceText documentText, string languageId, ILspLogger logger)
+    {
+        var documentFilePath = uri.ParsedUri is not null ? ProtocolConversions.GetDocumentFilePathFromUri(uri.ParsedUri) : uri.UriString;
+
+        // https://github.com/dotnet/roslyn/issues/78421: MAS should be its own workspace
+        if (_metadataAsSourceFileService.TryAddDocumentToWorkspace(documentFilePath, documentText.Container, out var documentId))
+        {
+            var metadataWorkspace = _metadataAsSourceFileService.TryGetWorkspace();
+            Contract.ThrowIfNull(metadataWorkspace);
+            var document = metadataWorkspace.CurrentSolution.GetRequiredDocument(documentId);
+            return document;
+        }
+
+        var primordialDoc = AddPrimordialDocument(uri, documentText, languageId);
+
+        if (uri.ParsedUri?.IsFile is true
+            && primordialDoc.Project.Language == LanguageNames.CSharp
+            && GlobalOptionService.GetOption(LanguageServerProjectSystemOptionsStorage.EnableFileBasedPrograms))
+        {
+            Contract.ThrowIfNull(primordialDoc.FilePath);
+            await BeginLoadingProjectAsync(primordialDoc.FilePath, primordialProjectId: primordialDoc.Project.Id);
+        }
+
+        return primordialDoc;
+
+        TextDocument AddPrimordialDocument(DocumentUri uri, SourceText documentText, string languageId)
+        {
+            var languageInfoProvider = _lspServices.GetRequiredService<ILanguageInfoProvider>();
+            if (!languageInfoProvider.TryGetLanguageInformation(uri, languageId, out var languageInformation))
+            {
+                Contract.Fail($"Could not find language information for {uri} with absolute path {documentFilePath}");
```

Comment:
I forgot that the caller needs to try/catch instead as part of this change. I think our higher level goal which is not fulfilled yet is that the 'Add' methods will return a non-null document or throw, and the LSP queue will catch exceptions, log, and move on, because it's important that the queue doesn't crash due to exceptions in any given request.

Reply:
do we intend to resolve this as part of this pr, or later on?

Reply:
I think we should just do it now. @jasonmalinowski @dibarbet do you have any thoughts on where a top-level try-catch for the LSP queue should be located?

Reply:
Maybe here?

https://github.com/dotnet/roslyn/blob/b610c623b7027522bd1015016674f1084acbc107/src/LanguageServer/Protocol/Handler/RequestContext.cs#L264-L283

Since I already see other logic reporting errors and catching the failure. That also generally covers us if the LSP solution forking logic were to go sideways too.

Reply:
And agreed on doing it now -- this is generally a reliability risk if we're assuming the misc files workspace would always succeed. It's easy to always succeed when it wasn't really doing anything before, but now if it is, we'll probably want this at some point.

Reply:
I dont think we want to catch everything there.  Just the misc files part.  If we can't apply the rest of the changes we definitely have an issue.  If we couldn't generate a misc document, that is OK (we throw if anyone actually asks for it later).

Reply:
> Just the misc files part.

I'm not following. Is this a suggestion to put the try/catch in a different location instead?

Reply:
@RikkiGibson just around these calls here then:

https://github.com/dotnet/roslyn/blob/7496b67cb1f86ef2da28f27af1d3809091cec5e1/src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs#L256-L257
https://github.com/dotnet/roslyn/blob/7496b67cb1f86ef2da28f27af1d3809091cec5e1/src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs#L273
https://github.com/dotnet/roslyn/blob/7496b67cb1f86ef2da28f27af1d3809091cec5e1/src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs#L156

Reply:
> If we can't apply the rest of the changes we definitely have an issue.

That's fair; any failure there means every request is probably going to blow up, so a restart would be best. Here, it's just affecting the misc files documents.

Reply:
Done